aioxmpp/bookmarks/xso.py

```diff
@@ -30,7 +31,96 @@
 namespaces.xep0048 = "storage:bookmarks"
 
 
-class Conference(xso.XSO):
+class Bookmark(xso.XSO):
+    """A bookmark XSO abstract base class.
+
+    Every XSO class registered as child of :class:`Storage` must be
+    a :class:`Bookmark` subclass.
+
+    Bookmarks must provide the following interface:
+
+    .. autoproperty:: primary
+
+    .. autoproperty:: secondary
+
+    .. autoproperty:: name
+
+    Equality is defined in terms of those properties:
+
+    .. automethod:: __eq__
+
+    It is highly recommenden not to redefine :meth:`__eq__` in a
+    subclass, if you do so make sure that the following law
+    holds for :meth:`__eq__`, :attr:`primary` and :attr:`secondary`:
+
+        (type(a) == type(b) and
+         a.primary == b.primary and
+         a.secondary == b.secondary)
+
+    if and only if
+
+        a == b
+
+    Otherwise the generation of bookmark change signals is not
+    guaranteed to be correct.
+    """
+
+    def __eq__(self, other):
+        """
+        Compare for equality by value and type.
+
+        The value of a bookmark must be fully determined by the values
+        of the :attr:`primary` and :attr:`secondary` properties.
+
+        This is used for generating the bookmark list change signals
+        and for the get-modify-set methods.
+        """
+        return (type(self) == type(other) and
+                self.primary == other.primary and
+                self.secondary == other.secondary)
+
+    @abstractproperty
+    def primary(self):
+        """
+        Return the primary category of the bookmark.
```

Comment:
No. I should document, that this is explicitly allowed (and that the inner structure of the primary category is opaque).